File: 09-语言学习/Golang/定时器与四叉堆/go/main.go
package main

import "fmt"

var checks, failed int

func check(cond bool, label string) {
	checks++
	if !cond {
		failed++
		fmt.Println("FAIL:", label)
	}
}

func checkEq(got, want interface{}, label string) {
	checks++
	if fmt.Sprint(got) != fmt.Sprint(want) {
		failed++
		fmt.Printf("FAIL: %s (got=%v want=%v)\n", label, got, want)
	}
}

func addAll(ts *Timers, whens []int64) []*Timer {
	var out []*Timer
	for i, w := range whens {
		t := &Timer{When: w, Name: fmt.Sprintf("t%d", i)}
		out = append(out, t)
		ts.AddHeap(t)
	}
	return out
}

func heapOK(ts *Timers) bool {
	for i := 1; i < len(ts.Heap); i++ {
		p := (i - 1) / timerHeapN
		if ts.Heap[p].When > ts.Heap[i].When {
			return false
		}
	}
	return true
}

func main() {
	// 1) 四叉堆拓扑：parent(i) = (i-1)/4
	ts := &Timers{}
	addAll(ts, []int64{50, 10, 90, 30, 70, 20, 80, 40})
	fmt.Println("heap:", func() []int64 {
		var xs []int64
		for _, tw := range ts.Heap {
			xs = append(xs, tw.When)
		}
		return xs
	}())
	checkEq(len(ts.Heap), 8, "8 个定时器入堆")
	checkEq(ts.Heap[0].When, int64(10), "堆顶是最小的 10")
	check(heapOK(ts), "堆序成立")
	for i := 1; i < len(ts.Heap); i++ {
		checkEq(ts.Heap[(i-1)/timerHeapN].When <= ts.Heap[i].When, true,
			fmt.Sprintf("heap[%d] 不小于其父", i))
	}

	// 2) Stop 只打标记
	head := ts.Heap[0].Timer
	checkEq(head.Stop(), true, "未触发时 Stop 返回 true")
	checkEq(len(ts.Heap), 8, "Stop 不从堆里摘除")
	checkEq(ts.Zombies, int32(1), "记 1 个 zombie")
	check(head.state&timerZombie != 0 && head.state&timerModified != 0, "Zombie+Modified 同时置位")
	checkEq(head.Stop(), false, "重复 Stop 返回 false")
	ts.CleanHead()
	checkEq(len(ts.Heap), 7, "cleanHead 摘掉堆顶 zombie")

	// 3) Reset 改晚：快照延后同步
	ts2 := &Timers{}
	rs := addAll(ts2, []int64{10, 20, 30})
	pending, _ := rs[0].Modify(25, 0)
	checkEq(pending, true, "Reset 的 pending 取自旧 when")
	checkEq(ts2.Heap[0].When, int64(10), "modify 不碰堆快照")
	ts2.UpdateHeap(rs[0])
	checkEq(ts2.Heap[0].When, int64(20), "updateHeap 后堆顶是 20")
	checkEq(ts2.Heap[1].When, int64(25), "被改晚的定时器下沉到下标 1")

	// 4) ticker 跳过 missed tick
	ts3 := &Timers{}
	tick := &Timer{When: 10, Period: 5, Name: "tick"}
	ts3.AddHeap(tick)
	checkEq(ts3.Run(9), int64(10), "now=9 返回下次时刻 10")
	checkEq(ts3.Run(27), int64(0), "now=27 触发")
	checkEq(tick.When, int64(30), "next = 10+5*(1+17/5) = 30")
	checkEq(ts3.Run(30), int64(0), "now=30 再触发")
	checkEq(tick.When, int64(35), "未延误时 next = 35")
	checkEq(len(ts3.Fired), 2, "共触发 2 次")

	// 5) 一次性定时器触发后被摘除（zombie 只是瞬时状态）
	ts4 := &Timers{}
	os := addAll(ts4, []int64{10, 20})
	checkEq(ts4.Run(10), int64(0), "一次性定时器触发")
	checkEq(ts4.ZombiePeak, int32(1), "触发瞬间曾置 zombie")
	checkEq(ts4.Zombies, int32(0), "updateHeap 又减回 0")
	check(os[0].ts == nil, "该定时器已离开堆")
	checkEq(len(ts4.Heap), 1, "堆里只剩一个")

	fmt.Printf("checks=%d failed=%d\n", checks, failed)
	if failed > 0 {
		panic("selfcheck failed")
	}
}

File: 09-语言学习/Golang/定时器与四叉堆/python/main.py
"""Go 定时器四叉堆：可运行演示。

模型逐行转写自官方源码 src/runtime/time.go，这里是四个能直接看出来的结论：
1) 堆是四叉的（timerHeapN=4），不是二叉；
2) Stop 只打 Zombie 标记，不从堆里摘除；
3) Reset 走 modify，堆里的 when 快照延后到 updateHeap/adjust 才同步；
4) ticker 用 next = when + period*(1+delay/period) 跳过 missed tick。
"""

from timermodel import Timer, Timers, TIMER_ZOMBIE, TIMER_MODIFIED


def section(title):
    print("\n== %s ==" % title)


def dump(ts):
    print("  heap = %s" % [(tw.timer.name, tw.when) for tw in ts.heap])


def main():
    section("1. 四叉堆：8 个定时器入堆后的父子关系")
    ts = Timers()
    for i, w in enumerate([50, 10, 90, 30, 70, 20, 80, 40]):
        ts.add_heap(Timer(w, name="t%d" % i))
    dump(ts)
    for i in range(1, len(ts.heap)):
        print("  heap[%d]=%d 的父节点 heap[%d]=%d"
              % (i, ts.heap[i].when, (i - 1) // 4, ts.heap[(i - 1) // 4].when))

    section("2. Stop 只打标记，堆长度不变")
    head = ts.heap[0].timer
    print("  Stop(%s) -> pending=%s" % (head.name, ts.stop(head)))
    dump(ts)
    print("  len(heap)=%d  zombies=%d  state(zombie=%s, modified=%s)"
          % (len(ts.heap), ts.zombies,
             bool(head.state & TIMER_ZOMBIE), bool(head.state & TIMER_MODIFIED)))
    ts.clean_head()
    dump(ts)
    print("  cleanHead 之后 len(heap)=%d（堆顶 zombie 走 deleteMin 摘除）" % len(ts.heap))

    section("3. Reset 改晚：快照延后同步")
    ts2 = Timers()
    timers = []
    for i, w in enumerate([10, 20, 30]):
        t = Timer(w, name="r%d" % i)
        timers.append(t)
        ts2.add_heap(t)
    print("  Reset(r0, 25) -> pending=%s wake=%s" % ts2.modify(timers[0], 25))
    dump(ts2)
    print("  ↑ 堆顶快照仍是 10，因为 modify 不碰堆")
    ts2.update_heap(timers[0])
    dump(ts2)
    print("  ↑ updateHeap 同步快照并 siftDown(0)")

    section("4. ticker 跳过 missed tick")
    ts3 = Timers()
    tick = Timer(10, period=5, name="tick")
    ts3.add_heap(tick)
    for now in (9, 27, 29, 30, 31):
        r = ts3.run(now)
        print("  run(now=%2d) -> %2d ; next when=%d ; fired=%d"
              % (now, r, tick.when, len(ts3.fired)))
    print("  delay=17 时 next=10+5*(1+17/5)=30，而不是 15/20/25 逐个补")


if __name__ == "__main__":
    main()
